### PR TITLE
add a webhook component for Kata Containers pods

### PR DESCRIPTION
**Pull request by liubin**
See https://github.com/kata-containers/kata-containers/issues/3976

### CODE REVIEW

**Comment by snir911** [comment]
nice! would it make sense to name it "annotations-webhook" or similar? as we also have the kata-webhook under the tests repo, ATM it's just a bit confusing but if we'll merge the repos it would be more confusing :p

**Comment by liubin** [comment]
@snir911 actually this webhook is based on the webhook in the `tests` repo, it seems that the webhook only adds the kata runtime class to Pod. Of course, we could combine the two into one, although I think that one is mainly for testing purposes

File: src/tools/kata-webhook/main.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+	if pod.Annotations == nil {
+		pod.Annotations = make(map[string]string)
+	}
+
+	annoationValue, err := eds.String()
+	if err != nil {
+		return nil, err
+	}
+	// pod.Annotations["io.katacontainers.sandbox.volumes.emptydir"] = annoationValue
```
Comment (amshinde) [inline src/tools/kata-webhook/main.go:84]: You may want to get rid of this commented code.

File: src/runtime/virtcontainers/kata_agent.go [hunk 2]
```diff
@@ -1341,7 +1341,7 @@ func (k *kataAgent) handleHugepages(mounts []specs.Mount, hugepageLimits []specs
 
 // handleEphemeralStorage handles ephemeral storages by
 // creating a Storage from corresponding source of the mount point
-func (k *kataAgent) handleEphemeralStorage(mounts []specs.Mount) ([]*grpc.Storage, error) {
+func (k *kataAgent) handleEphemeralStorage(mounts []specs.Mount, annoations map[string]string) ([]*grpc.Storage, error) {
```
Comment (amshinde) [inline src/runtime/virtcontainers/kata_agent.go:1344]: s/annoations/annotations/

**Comment by amshinde** [comment]
@liubin Approach looks good to me as it does solve the problem of respecting a sizeLimit provided by kubernetes, something that is missing today. 
But I do feel we should address this inherent issue in the long run at the CRI level by passing the sizeLimit information down the stack. I think there was some work being  done around this couple of years back, but I cant seem to find it. The only issue related to this that I had opened a while back about this : https://github.com/kata-containers/runtime/issues/2631

Would like to hear from others as well.

**Comment by egernst** [comment]
I have similar thoughts to what @amshinde already shared.

Ideally, we wouldn't need annotations at all, and if we did need/consume them, it'd be provided by the the CRI implementor via Kubelet when generating the config/spec, rather than via a web hook.

Do we have a list of the current limitations / gaps in support for features / parameters that aren't available to us from Kubelet/CRI? I feel like we're ad hoc working on each one to date (ie, see fsgroup support, direct volume support).

**Comment by arronwy** [comment]
Can one of the admins verify this patch?

File: src/tools/kata-webhook/main.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+	}
+
+	var port string
+
+	if cfg.TLSEnabled() {
+		port = ":443"
+		logger.Infof("Listening on %s", port)
+		err = http.ListenAndServeTLS(port, cfg.certFile, cfg.keyFile, whHandler)
+	} else {
+		port = ":8080"
```
Comment (jodh-intel) [inline src/tools/kata-webhook/main.go:148]: [security]: Do we really want to listen on all interfaces?

**Comment by fidencio** [comment]
@liubin, we're closing this PR as it's been opened and inactive for more than one year.
If you think this is a mistake and the PR should be re-opened, please, feel free to do so after rebasing it atop of current main.

**Pull request closed without merging**